README.md
wut
===
__wut - "Working Up Twitter":__ tool to collect and store data from the Twitter Stream API for analysis and statistical evaluation

Version 0.1.2

TOC:
====
* [Purpose](README.md#purpose)
* [System requirements](README.md#system-requirements)
* [Quick install](README.md#quick-install)
* [Future development](README.md#future-development)
* [Changelog](README.md#changelog)
* [License](README.md#license)

Purpose:
========
This application connects to the public streaming endpoint of the Twitter API (1.1) and collects tweets based on the search the user entered. It is able to track up to 400 keywords, 5000 Users and 25 geo locations defined through bounding boxes. (This limits are due to the restrictions of Twitter.)
This application was developed during the course of a bachelor thesis, which miserably failed, because the author lacks the ability of meeting deadlines.

System requirements:
===================
+ This application has been tested on a single core 2.60 GHz CPU with 2 gb of RAM on windowsXP with a 16 mBit/s downstream connection and is able to cope with about 55 Tweet messages per second (about 4 million tweets per day). The application itselfs needs between 20 and 80 mb of RAM depending on OS and incomming traffic. The database should have a greater portion of memory depending on your system setup and hard disk speed.
+ __OS:__ tested on windows XP and debian linux, should work on any OS that meets the other requirments
+ __Plattform:__ node.js >= v0.8.8 should run on previous versions but is not tested

    *__needed node modules:__*
    * [node-mysql](https://github.com/felixge/node-mysql)
    * [twit](https://github.com/ttezel/twit)

+ __Datbase:__ MySQL 5.5.3 or higher
    * To store the content of the tweets right you need to have utf8mb4 support. On older MySQL versions 4 bytes UTF-8 chars and emoji-codepoints are not stored correctly!

    * On Windows DO NOT install MySQL on a FAT-32 partiton or you will run into the 4 GB file limit of the FAT-32 filesystem very fast! Use NTFS on Windows instead.

    * If you expect high volume traffic from your Twitter tracks you may dedicate a whole partition to the database installation. Please see the MySQL manual on InnoDB setup.

    * You also might want to alter the standard MySQL configuration since 64mb Memory is a bit low for the database backend to handle higher volume traffic.

+ *Recommended:* a lot of diskspace on a fast hard disk and a nice chunk of RAM for the database! (Since it is the bottle neck of this application)

+ *Recommended:* For now a database frontend like phpMyAdmin, since a web frontend is not yet implemented. But through the database you can configure the searched tracks and access the collected data. (I use xampp, it comes with MySQL and phpMyAdmin and is easy to install on most systems - see: http://www.apachefriends.org)

+ *Recommended:* a fast internet connection with an adequate down stream. You do not need your own server. I devloped and tested this application from home with a cable connection (16 mBit/s = 2 MB/s downstream to connect to the sample stream endpoint of Twitter (peak: 55 msgs/s)). You should have at least a 165 kB/s down stream connection, or else you will be disconnected from the Twitter stream, because you are falling behind.

Quick install:
==============
* Install node.js (http://nodejs.org/)
* Install MySql (e.g. http://www.apachefriends.org)
* Download and unpack the source code to a directory of your choice
* Use npm in that directory to install the modules "node-mysql" and "twit" ("npm install"). This will put these required modules into the sub folder "node_modules"
* Create the needed tables with sql/CreateTables.sql and a MySQL-client of your choice
* Make the needed authentication keys on Twitter. Create twitConfig.js and dbConfig.js in the directory of the application. See: twitConfig.sample and dbConfig.sample as templates
* Insert your search terms into the tables "wut_follow", "wut_track" and "wut_locations"
* From the command prompt run "node twitterLoop.js"

Future development:
===================
Besides fixing bugs and improving performance, this application will get:
* Porper tests
* A proper web frontend, so the user, will not have to work directly through the database
* Language detection in real time
* Installation of the JavaScript core through npm

Changelog:
==========
__Version 0.1.2__

+ added table 'wut_retweets' and code to represent missing retweet->tweet relation

+ [changelog](changelog.md)

License:
========
(The MIT License)

Copyright (c) 2012 by Kai Koch kai.koch@gmail.com and contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


changelog.md
Version 0.1.2
===
+ added table wut_retweets and code to represent missing retweet->tweet relation

Version 0.1.1
===
+ DBwrapper updated to comply to node-mysql @2.0.0-alpha5 and above
+ style guide added to project

Version 0.1.0
===
+ initial codebase

classes/TwitWrapper.js
/*jslint vars: true, node: true, indent: 4, maxlen: 80 */
/*properties
    ID, addListenersToStream, affRows, affectedRows, assemble,
    autoReConnectTimmer, compareSearchParams, connect, connectStream, dbQueryCB,
    dbh, delete, deleteCount, dir, emitted, endtime, errCount, escape, exports,
    follow, getLastQuery, getTime, isArray, join, length, limit, limitCount,
    locations, log, message, name, newStuff, newStuffCount, on, onDelete,
    onError, onLimit, onScrubGeo, onStatusWithheld, onTweet, onUserWithheld,
    onWarning, prototype, push, query, queryFollow, queryLocations, queryTrack,
    reReadSearchParamsTimer, readSearchParams, reconStats, reconnectStream,
    resetTwitterStatistic, round, scrubGeoCount, scrub_geo, stall_warnings,
    starttime, statisticsTimer, statitics, statusWithheldCount, status_withheld,
    stop, stream, stringify, stuffToTrack, toISOString, totalProcessed, track,
    tweetCount, tweetMsg, twit, userWithheldCount, user_withheld,
    waitForStuffFromDb, warning, warningCount
*/
var deleteNotice = require('../functions/deleteNotice');
var isArray = require('util').isArray;
var limitNotice = require('../functions/limitNotice');
var parseError = require('../functions/parseError');
var scrubGeoNotice = require('../functions/scrubGeoNotice');
var statusWithheldNotice = require('../functions/statusWithheldNotice');
var tweet = require('../functions/tweet');
var userWithheldNotice = require('../functions/userWithheldNotice');
var warningNotice = require('../functions/warningNotice');

function TwitWrapper(daTwit, daDbh) {
    'use strict';
    var that = this;
    this.dbh = daDbh;
    this.twit = daTwit;
    this.stuffToTrack = {
        track: '',
        follow: '',
        locations: ''
    };
    this.newStuff = {};
    this.newStuffCount = 0;
    this.stream = null;
    this.starttime = null;
    this.readSearchParams();
    this.reReadSearchParamsTimer = setInterval(
        function () {that.readSearchParams(); },
        600000
    );
    this.autoReConnectTimmer = setInterval(
        function () {that.reconnectStream(); },
        21600000
    );
    this.statisticsTimer = setInterval(
        function () {that.statitics(); },
        60000
    );
}

TwitWrapper.prototype.addListenersToStream = function () {
    'use strict';
    var that = this;
    this.stream.on('connect', console.dir);
    this.stream.on('disconnect', console.dir);
    this.stream.on('reconnect', function (req, response, connectInterval) {
        that.reconStats(req, response, connectInterval);
    });
    this.stream.on('error', function (err) {
        that.onError(err);
    });
    this.stream.on('delete', function (msg) {
        that.onDelete(msg);
    });
    this.stream.on('limit', function (msg) {
        that.onLimit(msg);
    });
    this.stream.on('scrub_geo', function (msg) {
        that.onScrubGeo(msg);
    });
    this.stream.on('status_withheld', function (msg) {
        that.onStatusWithheld(msg);
    });
    this.stream.on('tweet', function (msg) {
        that.onTweet(msg);
    });
    this.stream.on('user_withheld', function (msg) {
        that.onUserWithheld(msg);
    });
    this.stream.on('warning', function (msg) {
        that.onWarning(msg);
    });
};

TwitWrapper.prototype.assemble = function (err, resultRows, whatToAssemble) {
    'use strict';
    if (err) {
        throw err;
    }
    var i, len;
    var rs = [];
    len = resultRows.length;
    for (i = 0; i < len; i += 1) {
        rs.push(resultRows[i][whatToAssemble]);
    }
    this.newStuff[whatToAssemble] = rs.join(',');
    this.waitForStuffFromDb();
};

TwitWrapper.prototype.compareSearchParams = function () {
    'use strict';
    var now;
    var stuffChanged = false;
    if (this.stuffToTrack.track !== this.newStuff.track) {
        this.stuffToTrack.track = this.newStuff.track;
        stuffChanged = true;
    }
    if (this.stuffToTrack.follow !== this.newStuff.follow) {
        this.stuffToTrack.follow = this.newStuff.follow;
        stuffChanged = true;
    }
    if (this.stuffToTrack.locations !== this.newStuff.locations) {
        this.stuffToTrack.locations = this.newStuff.locations;
        stuffChanged = true;
    }
    this.newStuff = {};
    if (stuffChanged) {
        if (this.stream) {
            this.stream.stop();
        }
        this.resetTwitterStatistic();
        this.connectStream();
    } else {
        now = new Date();
        console.log("[" + now.toISOString() + "] Search parameters have not" +
            " changed. Next lookup in 10 minutes.");
    }
};


TwitWrapper.prototype.connectStream = function () {
    'use strict';
    var now;
    var tracking = {stall_warnings: "true"};
    var hasStuff = false;
    if (this.stuffToTrack.track) {
        tracking.track = this.stuffToTrack.track;
        hasStuff = true;
    }
    if (this.stuffToTrack.follow) {
        tracking.follow = this.stuffToTrack.follow;
        hasStuff = true;
    }
    if (this.stuffToTrack.locations) {
        tracking.locations = this.stuffToTrack.locations;
        hasStuff = true;
    }
    if (hasStuff) {
        this.stream = this.twit.stream('statuses/filter', tracking);
        this.addListenersToStream();
    } else {
        now = new Date();
        console.log("[" + now.toISOString() + "] nothing to search for." +
            " I am having 10 minutes break.");
    }
};

TwitWrapper.prototype.dbQueryCB = function (err, result) {
    'use strict';
    var i, len;
    if (err) {
        if (err.message ===
                'Connection lost: The server closed the connection.') {
            console.log(this.dbh.getLastQuery());
            this.dbh.connect();
        } else {
            throw err;
        }
    } else {
        this.totalProcessed += 1;
        if (isArray(result)) {
            len = result.length;
            for (i = 0; i < len; i += 1) {
                this.affRows += result[i].affectedRows;
            }
        } else {
            this.affRows += result.affectedRows;
        }
    }
};

TwitWrapper.prototype.onDelete = function (msg) {
    'use strict';
    var sql = deleteNotice(msg["delete"], this.dbh.escape);
    var that = this;
    this.dbh.query(sql, function (err, result) {
        that.dbQueryCB(err, result);
    });
    this.deleteCount += 1;
    this.emitted += 1;
};

TwitWrapper.prototype.onError = function (err) {
    'use strict';
    var that = this;
    var sql = parseError('null', that.dbh.escape(new Date()),
        that.dbh.escape(err.tweetMsg), that.dbh.escape(err.name),
        that.dbh.escape(err.message));
    this.dbh.query(sql, function (err, result) {
        that.dbQueryCB(err, result);
    });
    console.dir(err);
    this.errCount += 1;
    this.emitted += 1;
};

TwitWrapper.prototype.onLimit = function (msg) {
    'use strict';
    var sql = limitNotice(msg.limit, this.dbh.escape);
    var that = this;
    this.dbh.query(sql, function (err, result) {
        that.dbQueryCB(err, result);
    });
    this.limitCount += 1;
    this.emitted += 1;
};

TwitWrapper.prototype.onScrubGeo = function (msg) {
    'use strict';
    var sql = scrubGeoNotice(msg.scrub_geo, this.dbh.escape);
    var that = this;
    this.dbh.query(sql, function (err, result) {
        that.dbQueryCB(err, result);
    });
    this.scrubGeoCount += 1;
    this.emitted += 1;
};

TwitWrapper.prototype.onStatusWithheld = function (msg) {
    'use strict';
    var sql = statusWithheldNotice(msg.status_withheld, this.dbh.escape);
    var that = this;
    this.dbh.query(sql, function (err, result) {
        that.dbQueryCB(err, result);
    });
    this.statusWithheldCount += 1;
    this.emitted += 1;
};

TwitWrapper.prototype.onTweet = function (msg) {
    'use strict';
    var sql = 'START TRANSACTION;' + tweet(msg, this.dbh.escape, false) +
              'COMMIT;';
    var that = this;
    this.dbh.query(sql, function (err, result) {
        that.dbQueryCB(err, result);
    });
    this.tweetCount += 1;
    this.emitted += 1;
};

TwitWrapper.prototype.onUserWithheld = function (msg) {
    'use strict';
    var sql = userWithheldNotice(msg.user_withheld, this.dbh.escape);
    var that = this;
    this.dbh.query(sql, function (err, result) {
        that.dbQueryCB(err, result);
    });
    this.userWithheldCount += 1;
    this.emitted += 1;
};

TwitWrapper.prototype.onWarning = function (msg) {
    'use strict';
    var sql = warningNotice(msg.warning, this.dbh.escape);
    var that = this;
    this.dbh.query(sql, function (err, result) {
        that.dbQueryCB(err, result);
    });
    this.warningCount += 1;
    this.emitted += 1;
};

TwitWrapper.prototype.queryFollow = function () {
    'use strict';
    var sql = 'SELECT DISTINCT `wut_follow`.`follow`' +
              ' FROM `wut_follow`' +
              ' WHERE `wut_follow`.`follow` <> ""' +
              ' AND `wut_follow`.`begin` <= NOW()' +
              ' AND `wut_follow`.`end` >= NOW()' +
              ' ORDER BY `wut_follow`.`follow` ASC LIMIT 0, 5000;';
    var that = this;
    this.dbh.query(sql, function (err, resultRows) {
        that.assemble(err, resultRows, "follow");
    });
};

TwitWrapper.prototype.queryLocations = function () {
    'use strict';
    var sql = 'SELECT DISTINCT CONCAT( SW,  ",", NE ) AS locations' +
              ' FROM `wut_locations`' +
              ' WHERE `wut_locations`.`begin` <= NOW()' +
              ' AND `wut_locations`.`end` >= NOW()' +
              ' ORDER BY `wut_locations`.`begin` ASC LIMIT 0 , 25';
    var that = this;
    this.dbh.query(sql, function (err, resultRows) {
        that.assemble(err, resultRows, "locations");
    });
};

TwitWrapper.prototype.queryTrack = function () {
    'use strict';
    var sql = 'SELECT DISTINCT `wut_track`.`track` FROM `wut_track`' +
              ' WHERE `wut_track`.`track` <> ""' +
              ' AND `wut_track`.`begin` <= NOW()' +
              ' AND `wut_track`.`end` >= NOW()' +
              ' ORDER BY `wut_track`.`track` ASC LIMIT 0, 400;';
    var that = this;
    this.dbh.query(sql, function (err, resultRows) {
        that.assemble(err, resultRows, "track");
    });
};

TwitWrapper.prototype.readSearchParams = function () {
    'use strict';
    this.queryFollow();
    this.queryLocations();
    this.queryTrack();
};

TwitWrapper.prototype.reconStats = function (req, response, connectInterval) {
    'use strict';
    var now = new Date();
    console.log("[" + now.toISOString() + "] conInterval" + connectInterval);
    console.dir(req);
    console.dir(response);
};

TwitWrapper.prototype.reconnectStream = function () {
    'use strict';
    this.stream.stop();
    this.resetTwitterStatistic();
    this.connectStream();
};

TwitWrapper.prototype.resetTwitterStatistic = function () {
    'use strict';
    var log = {};
    if (!this.starttime) {
        this.starttime = new Date();
    } else {
        this.endtime = new Date();
        log.ID = null;
        log.starttime = this.starttime;
        log.endtime = this.endtime;
        log.affRows = this.affRows;
        log.deleteCount = this.deleteCount;
        log.emitted = this.emitted;
        log.errCount = this.errCount;
        log.limitCount = this.limitCount;
        log.scrubGeoCount = this.scrubGeoCount;
        log.statusWithheldCount = this.statusWithheldCount;
        log.totalProcessed = this.totalProcessed;
        log.tweetCount = this.tweetCount;
        log.userWithheldCount = this.userWithheldCount;
        log.warningCount = this.warningCount;
        log.stuffToTrack = JSON.stringify(this.stuffToTrack);
        // Direct DB-Call for Object maping to key = 'value' in SET part.
        this.dbh.dbh.query('INSERT INTO wut_stream_statistics SET ?',
            log,
            function (err) {
                if (err) {
                    console.log(this.dbh.getLastQuery());
                    throw err;
                }
            });
        this.statitics();
        this.starttime = new Date();
    }
    this.affRows = 0;
    this.deleteCount = 0;
    this.emitted = 0;
    this.errCount = 0;
    this.limitCount = 0;
    this.scrubGeoCount = 0;
    this.statusWithheldCount = 0;
    this.totalProcessed = 0;
    this.tweetCount = 0;
    this.userWithheldCount = 0;
    this.warningCount = 0;
};

TwitWrapper.prototype.statitics = function () {
    'use strict';
    var now, elSec, readPerSecond, procPerSecond;
    var noticeMsg = '\n[%s] twitParserEmitted(%d)' +
        '\nincomingPerSecond(%d) processedPerSecond(%d) affectedRows(%d)' +
        '\ntotalProcessed(%d):' +
        '\ntweets(%d) errors(%d) delMsg(%d) limitMsgs(%d)' +
        '\nscrubGeo(%d) sWitheld(%d) uWitheld(%d) warnings(%d)\n';
    now = new Date();
    elSec = (now.getTime() - this.starttime.getTime()) / 1000;
    procPerSecond = Math.round((this.totalProcessed / elSec) * 1000) / 1000;
    readPerSecond = Math.round((this.emitted / elSec) * 1000) / 1000;
    console.log(noticeMsg,
        now.toISOString(),
        this.emitted,
        readPerSecond,
        procPerSecond,
        this.affRows,
        this.totalProcessed,
        this.tweetCount,
        this.errCount,
        this.deleteCount,
        this.limitCount,
        this.scrubGeoCount,
        this.statusWithheldCount,
        this.userWithheldCount,
        this.warningCount);
};

TwitWrapper.prototype.waitForStuffFromDb = function () {
    'use strict';
    this.newStuffCount += 1;
    if (this.newStuffCount < 3) {
        return;
    }
    this.newStuffCount = 0;
    this.compareSearchParams();
};

module.exports = TwitWrapper;